### PR TITLE
[python-package] drop Python 3.6 support, add Python 3.12

### PR DESCRIPTION
**Pull request by jameslamb**
Inspired by this discussion: https://github.com/microsoft/LightGBM/pull/6375#issuecomment-2018388889

Proposes the following:

* bump `python-requires` to  `>= 3.7`
* run the one "oldest versions of everything" job on Python 3.7 instead of Python 3.6
* run most CI jobs on 1 minor version of Python newer than what it's currently on
   - *which includes adding the first Python 3.12 jobs here!*
* narrowly pin dependencies in Python 3.8 jobs to a specific, known-working environment

### Why these versions?

I picked the oldest versions where `pip install` run in the `python:3.7` docker image found a wheel. Found that building those old versions `scipy` and `pandas` from their sdists failed with a variety of errors.

Summary:

* `cffi` = unchanged at `1.15.1` ([Jun 2022](https://pypi.org/project/cffi/1.15.1/))
* `numpy` = from `1.16.6` ([Dec 2019](https://pypi.org/project/numpy/1.16.6/)) to `1.19.1` ([Jul 2020](https://pypi.org/project/numpy/1.19.1/))
* `pandas` = from `0.24.0` ([Jan 2019](https://pypi.org/project/pandas/0.24.0/)) to `1.1.3` ([Oct 2020](https://pypi.org/project/pandas/1.1.3/))
* `pyarrow` = unchanged at `6.0.1` ([Nov 2021](https://pypi.org/project/pyarrow/6.0.1/))
* `scikit-learn` = `0.18.2` ([Jun 2017](https://pypi.org/project/scikit-learn/0.18.2/)) to `0.24.0` ([Dec 2020](https://pypi.org/project/scikit-learn/0.24.0/))
* `scipy` = `0.19.0` ([Mar 2017](https://pypi.org/project/scipy/0.19.0/)) to `1.6.0` ([Dec 2020](https://pypi.org/project/scipy/1.6.0/))

### CODE REVIEW

File: .github/workflows/cuda.yml [hunk 1]
```diff
@@ -30,17 +30,17 @@ jobs:
...
           - method: wheel
             compiler: gcc
-            python_version: "3.11"
+            python_version: "3.10"
             cuda_version: "11.8.0"
             task: cuda
           - method: source
             compiler: gcc
-            python_version: "3.9"
+            python_version: "3.12"
```
Comment (jameslamb) [inline .github/workflows/cuda.yml:38]: proposing swapping thise one and the `wheel` task, so we're always testing newest CUDA together with newest Python

**Review by borchero — APPROVED** [review]
LGTM!

**Comment by jameslamb** [comment]
Thanks for the review!

For breaking changes like this, I'd like to try to get 2 reviews. @jmoralez I know you'd supported this general idea in https://github.com/microsoft/LightGBM/pull/6375#issuecomment-2018388889, could you take a look at this PR and let us know what you think about the choices I made here?

**Review by jmoralez — APPROVED** [review]
Looks great. Thanks!

**Comment by jameslamb** [comment]
hmmmmm all the `gcc` CUDA jobs are failing with errors like this:

```text
__________________ ERROR collecting tests/c_api_test/test_.py __________________
tests/c_api_test/test_.py:10: in <module>
    from lightgbm.basic import _LIB as LIB
/github/home/.local/lib/python3.10/site-packages/lightgbm/__init__.py:8: in <module>
    from .basic import Booster, Dataset, Sequence, register_logger
/github/home/.local/lib/python3.10/site-packages/lightgbm/basic.py:278: in <module>
    _LIB = _load_lib()
/github/home/.local/lib/python3.10/site-packages/lightgbm/basic.py:262: in _load_lib
    lib = ctypes.cdll.LoadLibrary(lib_path[0])
/tmp/miniforge/envs/test-env/lib/python3.10/ctypes/__init__.py:452: in LoadLibrary
    return self._dlltype(name)
/tmp/miniforge/envs/test-env/lib/python3.10/ctypes/__init__.py:374: in __init__
    self._handle = _dlopen(self._name, mode)
E   OSError: libomp.so.5: cannot open shared object file: No such file or directory
```

([example build link](https://github.com/microsoft/LightGBM/actions/runs/9201567332/job/25359321703?pr=6459))

I see this in the CMake logs:

```text
-- Found OpenMP_C: -fopenmp (found version "4.5")
-- Found OpenMP_CXX: -fopenmp (found version "4.5")
-- Found OpenMP: TRUE (found version "4.5")
```

It's surprising to see LightGBM linking against `libomp`, not `libgomp`, using `gcc`. I have a feeling this is a result of building inside a `conda` environment but not using `conda`'s compilers (ref: https://github.com/microsoft/LightGBM/issues/5106#issuecomment-1121925896).

I'm gonna push a commit settings `LD_DEBUG=libs` so we can see more information.

**Comment by jameslamb** [comment]
I'm moving the testing of the CUDA CI issue to a separate PR, I don't think it's related to this PR's changes: #6465

**Comment by jameslamb** [comment]
I've merged in #6465, which fixes the CUDA CI issues, and reverted all the debugging stuff I'd pushed here.

Will merge this once CI (hopefully) passes.

**Pull request merged**